include/seastar/core/semaphore.hh

```diff
@@ -218,30 +276,75 @@ public:
     basic_semaphore(size_t count) noexcept(std::is_nothrow_default_constructible_v<exception_factory>)
         : exception_factory()
         , _state(make_lw_shared<state>(*this, count))
-        , _wait_list(expiry_handler{*this})
     {}
     basic_semaphore(size_t count, exception_factory&& factory) noexcept(std::is_nothrow_move_constructible_v<exception_factory>)
         : exception_factory(std::move(factory))
         , _state(make_lw_shared<state>(*this, count))
-        , _wait_list(expiry_handler{*this})
-    {
-        static_assert(std::is_nothrow_move_constructible_v<expiry_handler>);
-    }
+    {}
     basic_semaphore(basic_semaphore&& o) noexcept
         : exception_factory(std::move(o))
         , _state(std::move(o._state))
         , _wait_list(std::move(o._wait_list))
+        , _waiters_freelist(std::move(o._waiters_freelist))
     {
         if (_state->sem) {
             _state->sem = this;
         }
     }
-    basic_semaphore& operator=(basic_semaphore&&) = default;
+    basic_semaphore& operator=(basic_semaphore&& o) noexcept {
+        if (this != &o) {
+            broken();
+            exception_factory::operator=(std::move(o));
+            _state = std::move(o._state);
+            if (_state->sem) {
+                _state->sem = this;
+            }
+            _wait_list = std::move(o._wait_list);
+            _waiters_freelist.splice(_waiters_freelist.end(), o._waiters_freelist);
+        }
+        return *this;
+    }
 
     ~basic_semaphore() {
         broken();
     }
 
+private:
+    struct waiter : public basic_waiter {
+        using basic_waiter::basic_waiter;
+
+        promise<> pr;
+
+        future<> get_future() noexcept { return pr.get_future(); }
+        virtual void set_value() noexcept override {
+            pr.set_value();
+            // note: we self-delete in either case we are woken
+            // up. See usage below: only the resulting future
+            // state is required once we've left the _wait_list
+            delete this;
+        }
+        virtual void set_exception(std::exception_ptr ex) noexcept override {
+            pr.set_exception(std::move(ex));
+            // note: we self-delete in either case we are woken
+            // up. See usage below: only the resulting future
+            // state is required once we've left the _wait_list
+            delete this;
+        }
+    };
+
+    boost::intrusive::list<waiter, boost::intrusive::constant_time_size<false>> _waiters_freelist;
```

Comment:
shall we call `_waiters_freelist.clear_and_dispose(std::default_delete<waiter>{});` in the dtor?

Reply:
I think that's done auto-magically when the boost::intrusive::list is destroyed.

Reply:
That's tested in test_destroy_semaphore_during_wait_with_ensured_space

Reply:
IIUC, unlike STL containers which use the value semantics, boost::intrusive::list considers the elements as nodes, and does not "own" the elements in the sense that their dtor get called in the list's dtor. see https://godbolt.org/z/19K6PG4e1

Reply:
Sorry, you're right.
This is actually already solved in https://github.com/scylladb/seastar/pull/1477/commits/ff6df456112481415f66c2e8b9c14fe76911226b
in `broken()` (that's called in the dtor)

Reply:
i must be missing something, `broken()` is provided by the parent exception factory, which is by default `semaphore_default_exception_factory`, so `semaphore_default_exception_factory::broken()` creates an instance of `broken_semaphore`. i don't understand why this could help to avoid the leak.

Reply:
This `broken()`:
https://github.com/scylladb/seastar/blob/c0bf5768337dcec5e764700a4bcade33cca731e3/include/seastar/core/semaphore.hh#L468